posts/frangeris/04.upgradeable-contracts.md

```diff
@@ -0,0 +1,228 @@
+
+Here is an example of how `delegatecall` works:
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+// NOTE: Deploy this contract first
+contract B {
+    // NOTE: storage layout must be the same as contract A
+    uint public num;
+    address public sender;
+    uint public value;
+
+    function setVars(uint _num) public payable {
+        num = _num;
+        sender = msg.sender;
+        value = msg.value;
+    }
+}
+
+contract A {
+    uint public num;
+    address public sender;
+    uint public value;
+
+    function setVars(address _contract, uint _num) public payable {
+        // A's storage is update, B state is not modified.
+        (bool success, bytes memory data) = _contract.delegatecall(
+            abi.encodeWithSignature("setVars(uint256)", _num)
+        );
+    }
+}
+```
+
+In short words, the proxy contract is the one that DApps interact with, while using the logic of other contract in the proxy context. This means that the proxy contract will be the one that is deployed and the one that users will interact with, while the logic contract will be the one that is upgraded.
+
+![Meme about how proxy is behind scenes](./assets/04.image.1.png)
+
+More technical documentation can be found on the [Solidity docs](https://docs.soliditylang.org/en/v0.8.20/introduction-to-smart-contracts.html#delegatecall-and-libraries).
+
+> When working with upgradeable contracts using OpenZeppelin Upgrades, there are a few minor caveats to keep in mind when writing your Solidity code.
+>
+> It’s worth mentioning that these restrictions have their roots in how the Ethereum VM works, and apply to all projects that work with upgradeable contracts, not just OpenZeppelin Upgrades.
+
+Some of the caveats of using the proxy pattern are:
+
+1. Initializers: Contructors are not allowed in upgradeable contracts, instead we need to use the `initialize` function from the `Initializable` contract. This function is called only once when the contract is deployed for the first time. This means that we need to use the `initializer` modifier in the `initialize` function.
+
+2. Extending from upgradeable contracts libraries: Following the same logic, we need to use the `initializer` modifier when extending from upgradeable contracts libraries for libraries that we are extending from.
+
+3. Initial values and fields declarations: We need to be careful when declaring fields in our contracts, because the initial values are not set when the contract is deployed. This means that we need to set the initial values in the `initialize` function.
+
+4. Initialization of implemented contracts: As constructors are disabled when working with upgradeability, while extending from a contract, we call the function that "initialize" instead of calling the parent via `super`.
```

Comment:
```suggestion
4. Initialization of implemented contracts: As constructors are disabled when working with upgradeability, we call the function that "initializes" the contract instead of the parent via `super` while extending from another one.
```

Reply:
I think _the contract_ is the thing being initialized; if not, sth else should be used after _initialize_.

Reply:
Not sure about this one, I want to emphasize the `super` issue